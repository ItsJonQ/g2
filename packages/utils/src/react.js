import { Children, isValidElement } from 'react';

/**
 * Merges React `ref` together.
 */
export { default as mergeRefs } from 'react-merge-refs';

/**
 * Copies non-react specific statics from a child component to a parent component.
 * Similar to Object.assign, but with React static keywords prevented from being overridden.
 */
export { default as hoistNonReactStatics } from 'hoist-non-react-statics';

/**
 * Gets a collection of available children elements from a React component's children prop.
 *
 * @param {import('react').ReactNode} children
 *
 * @return {import('react').ReactNodeArray} An array of available children.
 */
export function getValidChildren(children) {
	if (typeof children === 'string') return [children];

	return Children.toArray(children).filter((child) => isValidElement(child));
}

/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}

/**
 * Checks to see if a `children` prop is a render function prop.
 *
 * @param {*} children
 *
 * @return {boolean} True, if children is a render function prop.
 */
export function isRenderProp(children) {
	return typeof children === 'function';
}

/**
 * Handles the rendering of a React component's children prop, which
 * may be a render function prop.
 *
 * @param {*} children Children to render.
 * @param {*} props Props to pass into a (potential) children render function.
 *
 * @return {*} The rendered children.
 */
export function renderChildren(children, props = {}) {
	if (isRenderProp(children)) {
		// eslint-disable-next-line no-unused-vars
		const { children: _, ...rest } = props;
		return children(rest);
	}

	return children;
}
