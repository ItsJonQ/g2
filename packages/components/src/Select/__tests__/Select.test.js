import { render } from '@testing-library/react';
import React from 'react';

import { Select } from '../index';

describe('props', () => {
	const options = [
		{ value: 'olaf', label: 'Olaf' },
		{ value: 'samantha', label: 'Samantha' },
	];

	test('should render correctly', () => {
		const { container } = render(<Select id="select" options={options} />);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render children', () => {
		const { container } = render(
			<Select id="select">
				<optgroup label="Snowmen">
					<option value="olaf">Olaf</option>
				</optgroup>
			</Select>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render disabled', () => {
		const { container } = render(
			<Select disabled id="select" options={options} />,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render required', () => {
		const { container } = render(
			<Select id="select" options={options} required />,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render readOnly', () => {
		const { container } = render(
			<Select id="select" options={options} readOnly />,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render size', () => {
		const { container } = render(
			<Select id="select" options={options} size="small" />,
		);
		expect(container.firstChild).toMatchSnapshot();
	});
});
