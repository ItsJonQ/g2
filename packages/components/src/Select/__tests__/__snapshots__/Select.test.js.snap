// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render children 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
>
  <select
    class="css-pjq1vj"
    data-g2-c16t="true"
    data-g2-component="Select"
    id="select"
  >
    <optgroup
      label="Snowmen"
    >
      <option
        value="olaf"
      >
        Olaf
      </option>
    </optgroup>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
>
  <select
    class="css-pjq1vj"
    data-g2-c16t="true"
    data-g2-component="Select"
    id="select"
  >
    <option
      value="olaf"
    >
      Olaf
    </option>
    <option
      value="samantha"
    >
      Samantha
    </option>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;

exports[`props should render disabled 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
  disabled=""
>
  <select
    class="css-pjq1vj"
    data-g2-c16t="true"
    data-g2-component="Select"
    disabled=""
    id="select"
  >
    <option
      value="olaf"
    >
      Olaf
    </option>
    <option
      value="samantha"
    >
      Samantha
    </option>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;

exports[`props should render readOnly 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
>
  <select
    class="css-pjq1vj"
    data-g2-c16t="true"
    data-g2-component="Select"
    id="select"
    readonly=""
  >
    <option
      value="olaf"
    >
      Olaf
    </option>
    <option
      value="samantha"
    >
      Samantha
    </option>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;

exports[`props should render required 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
>
  <select
    class="css-pjq1vj"
    data-g2-c16t="true"
    data-g2-component="Select"
    id="select"
    required=""
  >
    <option
      value="olaf"
    >
      Olaf
    </option>
    <option
      value="samantha"
    >
      Samantha
    </option>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;

exports[`props should render size 1`] = `
<div
  class="components-flex wp-components-flex components-base-field wp-components-base-field components-select wp-components-select css-1nl6bkg"
  data-g2-c16t="true"
  data-g2-component="SelectWrapper"
>
  <select
    class="css-1kjecn5"
    data-g2-c16t="true"
    data-g2-component="Select"
    id="select"
  >
    <option
      value="olaf"
    >
      Olaf
    </option>
    <option
      value="samantha"
    >
      Samantha
    </option>
  </select>
  <div
    class="components-flex-item wp-components-flex-item css-5yrsrk"
    data-g2-c16t="true"
    data-g2-component="FlexItem"
  >
    <span
      class="components-truncate wp-components-truncate components-text wp-components-text css-1fdlgmt"
      data-g2-c16t="true"
      data-g2-component="Text"
    >
      <div
        class="components-icon wp-components-icon css-1tw4gp0"
        data-g2-c16t="true"
        data-g2-component="Icon"
      >
        <svg
          fill="none"
          height="14"
          size="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </span>
  </div>
</div>
`;
